README.md
Device::Osram::Lightify
=======================

This module allows an Osram Lightify system to be controlled by Pure
Perl.

It is assumed that you will have paired your hub with all your bulbs,
and otherwise completed the setup of your system, via the mobile
application, before you use this code.

This code allows you to retrieve all the known-lights, by querying
your local hub, and apply operations to them.


# Current Status

* Introspection works such that you can dynamically retrieve all known
bulbs and their state:
     * On vs. Off
     * Brightness
     * Temperature
     * MAC address
     * Firmware Version
     * Name.
     * R,G,B,W values.
* Sending "all on" works.
* Sending "all off" works.
* You can set a specific light on, or off.
* You can change the brightness of a specific bulb.
* You can change the R,G,B,W values of a specific bulb.
* You can change the temperature of a specific bulb.

This is sufficient to make this a useful module.


# Sample Code

These examples all talk to the hub at IP address `192.168.10.136`,
you'd obviously specify the IP to your own hub there!

Turn all devices off:

    #!/usr/bin/perl

    use Device::Osram::Lightify::Hub;

    my $x = Device::Osram::Lightify::Hub->new( host => "192.168.10.136" );
    $x->all_off();

Or turn on only the light with name `hall`:

    #!/usr/bin/perl
    use Device::Osram::Lightify::Hub;

    my $x = Device::Osram::Lightify::Hub->new( host => "192.168.10.136" );

    foreach my $light ( $x->lights() ) {
        if ( $light->name() eq "hall" ) {
            $light->set_on();
            exit(0);
        }
    }

# Sample Application

There is a sample application included with the module, which lets you
carry out basic operations:

    $ ol --hub 192.168.10.136 --all-on
    $ ol --hub 192.168.10.136 --all-off
    $ ol --hub 192.168.10.136 --list
    Name: kitchen
        MAC:8418260000d9c70c
        version:1.2.4.1
        Brightness:100
        RGBW:255,255,255,255
        Temperature:2702
        Status:off
    Name: hall
        MAC:8418260000cb433b
        version:1.2.4.1
        Brightness:100
        RGBW:255,255,255,255
        Temperature:2702
        Status:on
    $ ol --hub 192.168.10.136 --off=hall



Steve
--


ChangeLog

0.1 - 2016/01/01
  * Initial release of the module to CPAN.
  * Sending "all on" and "all off" events works.
  * Detecting and displaying details about light-bulbs works.

0.2 - 2016/01/01
  * Controlling a bulb individually works:
	* Turning on.
	* Turning off.
	* Setting brightness.
	* Setting RGBW values.
        * Setting the temperature.

0.3 - 2016/01/01
  * Repacking update to allow CPAN to index.
    * Zero code-changes.

0.4 - 2016/01/02
  * Significant updates to the POD/documentation.
    * Zero code-changes.

0.5 - 2016/01/09
  * Added module beneath Power::Outlet namespace.

0.6 - 2016/10/01
  * Minor documentation updates for the Power::Output support.

0.7 - 2017/10/28
  * Added `--help` support to `ol`.
  * Ensure `--host` is supplied to `ol`.
  * Make setters chainable, implemented by Tekki <http://tekki.ch>
  * Decode light-names from UTF-8, implemented by Tekki <http://tekki.ch>


Makefile.PL

use strict;

use ExtUtils::MakeMaker 6.48;


my %WriteMakefileArgs = (
    NAME           => 'Device::Osram::Lightify',
    VERSION_FROM   => 'lib/Device/Osram/Lightify/Version.pm',
    EXE_FILES      => ["bin/ol" ],

    BUILD_REQUIRES => {

        # solely for the test-suite.
        'Test::More'   => 0,
        'Test::Pod'    => 0,
        'Test::Strict' => 0,
        'Test::NoTabs' => 0,

    },
    ABSTRACT         => 'Interface to the Osram Lightify system.',
    AUTHOR           => 'Steve Kemp <steve@steve.org.uk>',
    LICENSE          => 'perl',
    MIN_PERL_VERSION => '5.008',
    META_MERGE       => {
              resources => {
                  license    => 'http://dev.perl.org/licenses/',
                  homepage   => 'https://github.com/skx/Device-Osram-Lightify/',
                  bugtracker => 'https://github.com/skx/Device-Osram-Lightify/issues',
                  repository => 'https://github.com/skx/Device-Osram-Lightify.git',
              },
    },
);


WriteMakefile(%WriteMakefileArgs);


package MY;

sub metafile
{
    my $self = shift;
    my $meta = $self->SUPER::metafile_target(@_);
    my ($extra) = $self->echo( 'license: perl_5', 'MYMETA.yml', 1 );
    return "$meta\t$extra\n";
}


lib/Power/Outlet/Osram.pm
package Power::Outlet::Osram;

use strict;
use warnings;

use Device::Osram::Lightify::Hub;

BEGIN
{

    my $base = "use base qw{Power::Outlet::Common::IP};";
    eval($base);
    if ($@)
    {
        warn "Base-class Power::Outlet::Common::IP not found.\n";
    }
}

our $VERSION = '0.5';

=head1 NAME

Power::Outlet::Osram - Control and query an Osram Lightify light

=head1 SYNOPSIS

Using power-outlet shell script from Power::Outlet distribution

  power-outlet Osram ON host 192.168.1.10 name hall

Using Power::Outlet API

  my $outlet=Power::Outlet->new(type=>"Osram", host => "192.168.1.10", name=>"hall");
  print $outlet->on, "\n";

Using Power::Outlet::Osram  directly

  my $outlet=Power::Outlet::Osram->new(host => "192.168.1.10", name=>"hall");
  print $outlet->query, "\n";
  print $outlet->on, "\n";
  print $outlet->off, "\n";

=head1 DESCRIPTION

Power::Outlet::Osram is a package for controlling and querying a light on an Osram Lightify network attached bridge.

=head1 USAGE

  use Power::Outlet::Osram;
  my $lamp=Power::Outlet::Osram->new(host=>"mybridge", name=>"hall");
  print $lamp->on, "\n";

=head1 CONSTRUCTOR

=head2 new

  my $outlet=Power::Outlet->new(type=>"Osram", host=>"192.168.10.136", name => "Hall" );
  my $outlet=Power::Outlet::Osram->new(host=>"mybridge", name="kitchen");

=head1 PROPERTIES

=head2 name

Name for the particular light as configured on the Osram Lightify bridge.

=cut

sub name
{
    my $self = shift;
    $self->{ "name" } = shift if @_;
    return $self->{ "name" };
}


=head1 METHODS

=head2 query

Return the current state of the specified device, as a string.

=cut

sub query
{
    my $self = shift;

    my $x = Device::Osram::Lightify::Hub->new( host => $self->host() );
    foreach my $device ( $x->lights() )
    {
        if ( $device->name() eq $self->name() )
        {
            return ( $device->stringify() );
        }
    }

}

=head2 on

Sends a message to the device to Turn Power ON

=cut

sub on
{
    my $self = shift;
    return $self->_call("on");
}

=head2 off

Sends a message to the device to Turn Power OFF

=cut

sub off
{
    my $self = shift;
    return $self->_call("off");
}


=head2 _call

Implementation method to send an on/off message to the given device.

=cut

sub _call
{
    my $self  = shift;
    my $state = shift;

    my $x = Device::Osram::Lightify::Hub->new( host => $self->host() );
    foreach my $device ( $x->lights() )
    {
        if ( $device->name() eq $self->name() )
        {
            if ( $state =~ /on/i )
            {
                $device->set_on();
                return 1;
            }
            if ( $state =~ /off/i )
            {
                $device->set_off();
                return 1;
            }
        }
    }
    return 0;
}

=head2 switch

Queries the device for the current status and then requests the opposite.

=cut

#see Power::Outlet::Common->switch

=head2 cycle

Sends messages to the device to Cycle Power (ON-OFF-ON or OFF-ON-OFF).

=cut

#see Power::Outlet::Common->cycle

=head1 AUTHOR

Steve Kemp <steve@steve.org.uk>

=cut

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2016 Steve Kemp <steve@steve.org.uk>.

This library is free software. You can modify and or distribute it under
the same terms as Perl itself.

=cut

1;


lib/Device/Osram/Lightify/API.pm

=head1 NAME

Device::Osram::Lightify::API - API Documentation

=head1 API Overview

The starting point for using this module is that you have a hub with
an IP address upon your local LAN, which is able to communicate successfully
with any installed bulbs you have.

The module itself is responsible for talking to your hub, and sending
commands to it to initiate actions, or perform introspection, so you'll
need the IP address of your hub available.

=cut

=head1 Overview

The communication to/from the hub is carried out via short binary
communication with the hub over TCP/IP on port 4000.

=cut

=head1 Light Discovery

To discover the lights which are available to your hub we connect and
send the following binary string to it:

=for example begin

    0x0B 0x00 0x00 0x13 0x00 0x00 0x00 0x00 0x01 0x00 0x00 0x00 0x00

=for example end

Once sent we then read back a header.  The header contains 11 bytes
of reply.  From this header we read the 10th byte which will tell us
the number of bulbs which are available.

Once we know the number of bulbs we can then read 50 bytes for each
bulb, and this block of data can be parsed to show the current state
of that specific bulb.

=for example begin

  01 - ID byte 1
  02 - ID byte 2
  03 - MAC Address 1
  04 - MAC Address 2
  05 - MAC Address 3
  05 - MAC Address 4
  06 - MAC Address 5
  07 - MAC Address 6
  08 - MAC Address 7
  09 - MAC Address 8
  10 - Bulb Type
  11 - Firmware Version 1
  12 - Firmware Version 2
  13 - Firmware Version 3
  14 - Firmware Version 4
  15 - Online/Offline
  16 - Group ID 1
  17 - Group ID 2
  18 - Status 0 == off, 1 == on
  19 - Brightness (0-100)
  20 - Temperature 1
  21 - Temperature 2
  22 - R
  23 - G
  24 - B
  25 - W
  26 - Name 1
  27 - Name 2
  28 - Name 3
  29 - Name 4
  30 - Name 5
  31 - Name 6
  .. - Name 15

=for example end

B<NOTE> The returned 50 bytes are NULL-terminated/padded.

You'll almost certainly want to make sure you can parse this stuff,
because users will want to operate upon bulbs by name - and the API
only allows you to operate on specific devices by MAC-address, so at
the very least you must be able to lookup `NAME -> MAC`.

=cut


=head1 Broadcast

There are two simple commands which will turn B<all> lights on, or off,
these are simple to get started with because you don't need to set the
MAC address of the bulb inside the command.

To turn all bulbs on send this:

=for example begin

    0x0f 0x00 0x00 0x32 0x01 0x00 0x00 0x00 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0x01

=for example end

To turn all bulbs off send this:

=for example begin

    0x0f 0x00 0x00 0x32 0x01 0x00 0x00 0x00 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0x00

=for example end

Once sent read back 20 bytes to get your result.

=cut


=head1 Light-Specific Commands

The rest of the commands are going to be bulb-specific, and involve
sending a command with the MAC address of the destination device inside
their bodies.

It seems to be the case that we need to set a session ID with each
command, to avoid issues with commands being dropped as previously-seen
or otherwise prevent replay-attacks.

I've denoted such bytes as SS SS SS SS in the following commands.

=cut

=head2 On/Off

To set a specific bulb ON or OFF you need to send the following
magic string:

=for example begin

    0F 00 00 32 SS SS SS SS NN NN NN NN NN NN NN NN ON

=for example end

Here `NN` should be replaced with the MAC address of the device you
wish to control, backwards, and the last byte C<ON> should be replaced
by C<0x00> to turn the device off, and C<0x01> to turn it on.

Once set read back 20 bytes to get your result.

=cut

=head2 Brightness

To set a specific brightness for a bulb you need to send the following
magic string:

=for example begin

     11 00 00 31 SS SS SS SS NN NN NN NN NN NN NN NN XX 00 00

=for example end

Here `NN` should be replaced with the MAC address of the device you
wish to control, backwards, and the byte C<XX> should be replaced
with the brighness level you wish to set, in the range 0-100.

Once sent read back 20 bytes to get your result.

=cut

=head2 Colour

To set specific RGBW values for a bulb you need to send the following
magic string:

=for example begin

 14 00 00 36 SS SS SS SS NN NN NN NN NN NN NN NN RR GG BB WW 00 00

=for example end

Here `NN` should be replaced with the MAC address of the device you
wish to control, backwards, and the bytes RR,GG,BB,WW should be updated
with the appropriate values.

Once sent read back 20 bytes to get your result.

=cut


use strict;
use warnings;

1;


lib/Device/Osram/Lightify/Hub.pm

=head1 NAME

Device::Osram::Lightify::Hub - Communicate with an Osram Lightify Hub

=head1 DESCRIPTION

This module allows basic operation of the Osram lightify bulbs,
via connections to the Osram hub.

=cut

=head1 SYNOPSIS

   use Device::Osram::Lightify;

   my $tmp = Device::Osram::Lightify::Hub->new( host => "1.2.3.4" );

   # Turn all devices on
   $tmp->all_on();

   # Turn all devices off
   $tmp->all_of();

=cut

=head1 DESCRIPTION

This module will connect to an Osram Lightify hub, allowing the
control of lights via Perl.

The communication with the hub is carried out by sending/receiving
binary messages to the hub on port 4000.  The specific bytes set
have been documented in the L<Osram::Lightify::API> module.

This module is responsible for:

=over 8

=item Identifying lights:

We can initiate a discovery of all the available lights, and create
a suitable L<Osram::Lightify::Light> object for each discovered light.

=item Sending broadcast events:

We can send a broadcast event, which applies to all known lights,
to instruct them to go on or off.

=back

=cut

=head1 METHODS

=cut

use strict;
use warnings;

package Device::Osram::Lightify::Hub;

use IO::Socket::INET;
use Device::Osram::Lightify::Light;


=head2 new

Create a new hub-object, it is mandatory to provide a C<host> parameter
which will give the IP (and optional port) of the Osram hub.

=cut

sub new
{
    my ( $proto, %supplied ) = (@_);
    my $class = ref($proto) || $proto;

    my $self = {};
    bless( $self, $class );


    $self->{ 'host' } = $supplied{ 'host' } || die "Missing host parameter";

    return $self;
}


=head2 all_on

Broadcast an "on" event to all lights.

=cut

sub all_on
{
    my ($self) = (@_);

    # Get the open socket
    $self->_connect() unless ( $self->{ '_socket' } );
    my $sock = $self->{ '_socket' };

    # Send the magic to initiate "All On"
    my $x = "";
    foreach my $char (
        qw! 0x0f 0x00 0x00 0x32 0x01 0x00 0x00 0x00 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0x01 !
      )
    {
        $x .= chr( hex($char) );
    }
    syswrite( $sock, $x, length($x) );

    # Read 8-byte header + 12-byte reply
    my $buffer = $self->_read(20);
    die 'Unable to turn all lights on!' if $self->_check_response($buffer);
}


=head2 all_off

Broadcast an "off" event to all lights.

=cut

sub all_off
{
    my ($self) = (@_);

    # Get the open socket
    $self->_connect() unless ( $self->{ '_socket' } );
    my $sock = $self->{ '_socket' };

    # Send the magic to initiate "All Off"
    my $x = "";
    foreach my $char (
        qw! 0x0f 0x00 0x00 0x32 0x01 0x00 0x00 0x00 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0x00 !
      )
    {
        $x .= chr( hex($char) );
    }
    syswrite( $sock, $x, length($x) );

    # Read 8-byte header + 12-byte reply
    my $buffer = $self->_read(20);
    die 'Unable to turn all lights off!' if $self->_check_response($buffer);
}



=head2 lights

Return a new C<Osram::Lightify::Light> object for each of the lights
that could be discovered.

=cut

sub lights
{
    my ($self) = (@_);

    my @ret;

    # Get the open socket
    $self->_connect() unless ( $self->{ '_socket' } );
    my $sock = $self->{ '_socket' };

    # Send the magic to initiate a scan.
    my $x = "";
    foreach my $char (
         qw! 0x0B 0x00 0x00 0x13 0x00 0x00 0x00 0x00 0x01 0x00 0x00 0x00 0x00 !)
    {
        $x .= chr( hex($char) );
    }
    syswrite( $sock, $x, length($x) );

    # Read 8-byte header + 3 bytes reply
    my $buffer = $self->_read(11);

    # Eight byte header we ignore.
    # 0 = ??
    # 1 = ??
    # ... ??
    # 8 = ??
    # 9 = Number of bulbs
    # 10 = ??

    # The number of devices.
    my $count = ord( substr( $buffer, 9, 1 ) );

    # For each one.
    while ($count)
    {
        # Read 8 byte header + 42 bytes for each light.
        my $buffer = $self->_read(50);
        $count = $count - 1;

        push( @ret,
              Device::Osram::Lightify::Light->new( hub    => $self,
                                                   binary => $buffer
                                                 ) );
    }

    return (@ret);
}


=begin doc _check_response

Internal method used in L<Hub|Osram::Lightify::Hub> and L<Light|Osram::Lightify::Light>.

Checks the response of the hub and returns the error code or 0 for success.

=end doc

=cut

sub _check_response()
{
    my ($self, $response) = (@_);

    $response =~ /(.)$/;
    return unpack('C', $1) * 1;
}


=begin doc _connect

Private and internal-method.

Connect to the hub, via the C<host> parameter we were given in our constructor.

=end doc

=cut

sub _connect
{
    my ($self) = (@_);

    my $host = $self->{ 'host' };
    my $port = 4000;

    if ( $host =~ /^(.*):([0-9]+)$/ )
    {
        $host = $1;
        $port = $2;
    }

    $self->{ '_socket' } =
      IO::Socket::INET->new( Proto    => "tcp",
                             Type     => SOCK_STREAM,
                             Blocking => 1,
                             PeerAddr => $host,
                             PeerPort => $port,
                           );

    die "Failed to connect to $host:$port" unless ( $self->{ '_socket' } );
    binmode( $self->{ '_socket' } );
}


=begin doc

Private and internal-method.

Read N-bytes from the open socket.  We do this a byte at a time,
and return only when we've read as much as we should.

=end doc

=cut

sub _read
{
    my ( $self, $count ) = (@_);

    my $out;

    while ($count)
    {
        my $buf;
        my $c = sysread( $self->{ '_socket' }, $buf, 1, 0 );
        if ($c)
        {
            $count -= $c;
            $out .= $buf;

        }

    }
    return ($out);
}


=begin doc _session_token

Private and internal-method.

Generate and return a four-byte token.  All the mutator commands
for the light-devices take such a thing.  It doesn't seem to matter
what we set, so I've encoded the seconds-past-the-epoch.

Setting a value avoids commands being dropped as "old"/"reused".

=end doc

=cut

sub _session_token()
{
    my $time = time();
    $time = sprintf( "%x", time );

    my $t = "";

    if ( $time =~ /^(..)(..)(..)(..)$/ )
    {
        $t .= chr( hex($1) );
        $t .= chr( hex($2) );
        $t .= chr( hex($3) );
        $t .= chr( hex($4) );
    }
    else
    {

        $t .= chr( hex("0x00") );
        $t .= chr( hex("0x00") );
        $t .= chr( hex("0x00") );
        $t .= chr( hex("0x00") );
    }
    return ($t);
}


1;



=head1 AUTHOR

Steve Kemp <steve@steve.org.uk>

=cut

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2016 Steve Kemp <steve@steve.org.uk>.

This library is free software. You can modify and or distribute it under
the same terms as Perl itself.

=cut


lib/Device/Osram/Lightify/Light.pm

=head1 NAME

Device::Osram::Lightify::Light - The interface to a single light

=head1 DESCRIPTION

This module allows a single Osram lightify light to be manipulated.

Objects are not expected to be constructed manually, instead
they are discovered dynmically via communication with the hub.

=cut

=head1 SYNOPSIS

   use Device::Osram::Lightify;

   my $tmp = Device::Osram::Lightify::Hub->new( host => "1.2.3.4" );

   # Show all nodes we found
   # (Stringification means we dump all the state here.)
   foreach my $light ( $tmp->lights() ) {
      print $light;
   }

=cut

=head1 DESCRIPTION

This module allows basic control of an Osram Lightify light.

=cut

=head1 METHODS

=cut

use strict;
use warnings;

package Device::Osram::Lightify::Light;

use Encode 'decode';

#
#  Allow our object to treated as a string.
#
use overload fallback => 1, '""' => 'stringify';


=head2 new

Create a new light-object.

This is invoked by C<Hub:lights()> method, which will read a binary
string containing all the details of the light - we must then parse
it according to L<Device::Osram::Lightify::API>.

=cut

sub new
{
    my ( $proto, %supplied ) = (@_);
    my $class = ref($proto) || $proto;

    my $self = {};
    bless( $self, $class );


    $self->{ 'hub' } = $supplied{ 'hub' } || die "Missing 'hub' parameter";
    $self->{ 'binary' } = $supplied{ 'binary' } ||
      die "Missing 'binary' parameter";

    $self->_decode_binary();
    return $self;
}


=begin doc

Internal method, parse the status of a light.

=end doc

=cut

sub _decode_binary
{
    my ($self) = (@_);

    my $buffer = $self->{ 'binary' };

    #
    # my $str    = $buffer;
    # $str =~ s/(.)/sprintf("0x%x ",ord($1))/megs;
    # print "HEX:" . $str . "\n";
    #

    # Get the MAC
    my $mac = substr( $buffer, 2, 8 );
    foreach my $c ( reverse( split( //, $mac ) ) )
    {
        $self->{ 'mac' } .= sprintf( "%02x", ord($c) );
        $self->{ 'maddr' } .= $c;
    }
    $self->{ 'maddr' } = reverse( $self->{ 'maddr' } );

    # Get the firmware-version
    my $ver = substr( $buffer, 11, 4 );
    foreach my $c ( split( //, $ver ) )
    {
        $self->{ 'version' } .= "." if ( $self->{ 'version' } );
        $self->{ 'version' } .= sprintf( "%x", ord($c) );
    }

    if ( ord( substr( $buffer, 18, 1 ) ) eq 1 )
    {
        $self->{ 'status' } = "on";
    }
    else
    {
        $self->{ 'status' } = "off";
    }

    # Brightness
    $self->{ 'brightness' } = ord( substr( $buffer, 19, 1 ) );

    # Temperature in kelvins
    my $k1 = ord( substr( $buffer, 20, 1 ) );
    my $k2 = ord( substr( $buffer, 21, 1 ) );
    $self->{ 'temperature' } = ( $k1 + ( 256 * $k2 ) );

    # R,G,B,W
    $self->{ 'r' } = ord( substr( $buffer, 22, 1 ) );
    $self->{ 'g' } = ord( substr( $buffer, 23, 1 ) );
    $self->{ 'b' } = ord( substr( $buffer, 24, 1 ) );
    $self->{ 'w' } = ord( substr( $buffer, 25, 1 ) );

    # The name of the bulb.
    $self->{ 'name' } = decode 'UTF-8', substr( $buffer, 26, 15 );
    $self->{ 'name' } =~ s/\0//g;
}



=head2 brightness

Get the brightness value of this light (0-100).

=cut

sub brightness
{
    my ($self) = (@_);

    return ( $self->{ 'brightness' } );
}


=head2 mac

Get the MAC address of this light.

=cut

sub mac
{
    my ($self) = (@_);

    return ( $self->{ 'mac' } );
}


=head2 name

Return the name of this light.

=cut

sub name
{
    my ($self) = (@_);

    return ( $self->{ 'name' } );
}



=head2 rgbw

Return the current RGBW value of this light.

=cut

sub rgbw
{
    my ($self) = (@_);

    my $x = "";
    $x .= $self->{ 'r' };
    $x .= ",";
    $x .= $self->{ 'g' };
    $x .= ",";
    $x .= $self->{ 'b' };
    $x .= ",";
    $x .= $self->{ 'w' };

    return ($x);
}


=head2 status

Is the light C<on> or C<off> ?

=cut

sub status
{
    my ($self) = (@_);

    return ( $self->{ 'status' } );
}


=head2 temperature

Get the temperature value of this light (2200-6500).

=cut

sub temperature
{
    my ($self) = (@_);

    return ( $self->{ 'temperature' } );
}


=head2 version

Get the firmware version of this light.

=cut

sub version
{
    my ($self) = (@_);

    return ( $self->{ 'version' } );
}



=head2 set_on

Set this light to be "on".

=cut

sub set_on
{
    my ($self) = (@_);

    my $parent = $self->{ 'hub' };
    my $socket = $parent->{ '_socket' };

    # Prefix for sending a light on
    my $x = "";
    foreach my $char (qw! 0x0f 0x00 0x00 0x32 !)
    {
        $x .= chr( hex($char) );
    }

    # Add a time/session-token
    $x .= $parent->_session_token();

    # MAC address - binary - in reverse
    $x .= $self->{ 'maddr' };

    # Desired state: 1
    $x .= chr( hex("0x01") );

    syswrite( $socket, $x, length($x) );

    # Read 8-byte header + 12-byte reply
    my $buffer = $parent->_read(20);
    die 'Unable to turn light on!' if $parent->_check_response($buffer);

    $self->{ 'status' } = 'on';

    return $self;
}


=head2 set_off

Set this light to be "off".

=cut

sub set_off
{
    my ($self) = (@_);

    my $parent = $self->{ 'hub' };
    my $socket = $parent->{ '_socket' };

    # Prefix for sending a light off
    my $x = "";
    foreach my $char (qw! 0x0f 0x00 0x00 0x32 !)
    {
        $x .= chr( hex($char) );
    }

    # Add a time/session-token
    $x .= $parent->_session_token();

    # MAC address - binary - in reverse
    $x .= $self->{ 'maddr' };


    # Desired state: 0
    $x .= chr( hex("0x00") );

    syswrite( $socket, $x, length($x) );

    # Read 8-byte header + 12-byte reply
    my $buffer = $parent->_read(20);
    die 'Unable to turn light off!' if $parent->_check_response($buffer);

    $self->{ 'status' } = 'off';

    return $self;
}


=head2 set_brightness

Set the brightness value of this light - valid values are 0-100.

=cut

sub set_brightness
{
    my ( $self, $brightness ) = (@_);

    if ( $brightness < 0 )
    {
        $brightness = 0;
    }
    if ( $brightness > 100 )
    {
        $brightness = 100;
    }

    my $parent = $self->{ 'hub' };
    my $socket = $parent->{ '_socket' };

    # Prefix for changing the brightness.
    my $x = "";
    foreach my $char (qw! 11 00 00 31 !)
    {
        $x .= chr( hex($char) );
    }

    # Add a time/session-token
    $x .= $parent->_session_token();

    # MAC address - binary - in reverse
    $x .= $self->{ 'maddr' };

    # Desired brightness 0-100.
    $x .= chr($brightness);

    $x .= chr( hex("0x00") );
    $x .= chr( hex("0x00") );

    syswrite( $socket, $x, length($x) );

    # Read 8-byte header + 12-byte reply
    my $buffer = $parent->_read(20);
    die 'Unable to change brightness!' if $parent->_check_response($buffer);

    $self->{ 'brightness' } = $brightness;

    return $self;
}


=head2 set_rgbw

Set the specified RGBW values of this light.

=cut

sub set_rgbw
{
    my ( $self, $r, $g, $b, $w ) = (@_);

    my $parent = $self->{ 'hub' };
    my $socket = $parent->{ '_socket' };

    # Prefix for changing the RGBW values.
    my $x = "";
    foreach my $char (qw! 0x14 0x00 0x00 0x36 !)
    {
        $x .= chr( hex($char) );
    }


    # Add a time/session-token
    $x .= $parent->_session_token();

    # MAC address - binary - in reverse
    $x .= $self->{ 'maddr' };

    # The colours.
    $x .= chr($r);
    $x .= chr($g);
    $x .= chr($b);
    $x .= chr($w);

    # Two more bytes
    $x .= chr( hex("0x00") );
    $x .= chr( hex("0x00") );

    syswrite( $socket, $x, length($x) );

    # Read 8-byte header + 12-byte reply
    my $buffer = $parent->_read(20);
    die 'Unable to change RGBW!' if $parent->_check_response($buffer);

    $self->{ 'r' } = $r;
    $self->{ 'g' } = $g;
    $self->{ 'b' } = $b;
    $self->{ 'w' } = $w;

    return $self;
}


=head2 set_temperature

Set the specified temperature value for this light, in the range 2200-6500.

=cut

sub set_temperature
{
    my ( $self, $temp ) = (@_);

    my $parent = $self->{ 'hub' };
    my $socket = $parent->{ '_socket' };

    if ( $temp < 2200 )
    {
        $temp = 2200;
    }
    if ( $temp > 6500 )
    {
        $temp = 6500;
    }

    my $t1 = $temp % 256;
    my $t2 = ( $temp - $t1 ) / 256;

    my $x = "";
    foreach my $char (qw! 0x12 0x00 0x00 0x33 !)
    {
        $x .= chr( hex($char) );
    }

    # Add a time/session-token
    $x .= $parent->_session_token();

    # MAC address - binary - in reverse
    $x .= $self->{ 'maddr' };

    # The temperature.
    $x .= chr($t1);
    $x .= chr($t2);

    # Two more bytes
    $x .= chr( hex("0x00") );
    $x .= chr( hex("0x00") );

    syswrite( $socket, $x, length($x) );

    # Read 8-byte header + 12-byte reply
    my $buffer = $parent->_read(20);
    die 'Unable to change temperature!' if $parent->_check_response($buffer);

    $self->{ 'temperature' } = $temp;

    return $self;
}



=head2 stringify

Convert the record to a string, suitable for printing.

=cut

sub stringify
{
    my ($self) = (@_);
    my $txt = "";

    $txt .= "Name: " . $self->name() . "\n";
    $txt .= "\tMAC:" . $self->mac() . "\n";
    $txt .= "\tversion:" . $self->version() . "\n";
    $txt .= "\tBrightness:" . $self->brightness() . "\n";
    $txt .= "\tRGBW:" . $self->rgbw() . "\n";
    $txt .= "\tTemperature:" . $self->temperature() . "\n";
    $txt .= "\tStatus:" . $self->{ 'status' } . "\n";

    $txt;
}

1;



=head1 AUTHOR

Steve Kemp <steve@steve.org.uk>

=cut

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2016 Steve Kemp <steve@steve.org.uk>.

This library is free software. You can modify and or distribute it under
the same terms as Perl itself.

=cut
